### ISSUE TITLE
Javascript expressions possible improvements.

### ISSUE THREAD

**Issue by tomy2105** [elsa-workflows/elsa-core#807]
https://github.com/elsa-workflows/elsa-core/issues/785 has been closed but the general discussion that started in it should continue I think, resulting in some agreement on how to write activity outputs so they better integrate with JavaScript expressions.

So @craigfowler said:
> 
> I mostly-agree. To some extents though this is just the nature of what we are doing.
> ## I don't think it's a great idea to try and "be more like JavaScript"
> 
> Jint allows us to refer-to & manipulate objects from JavaScript language & syntax. But - Elsa is still fundamentally a .NET application, so a lot of the objects we are using came "from the .NET world". _JavaScript is the foreigner here and .NET objects are the natives_. 

Well, as a "hard core" programmer, same as you are, I agree with you and I probably could live with it, kind of.
However I envision that users using designer do not care what language Elsa is written in. 
They just care what they see in designer and designer says JavaScript.
Moreover, they could be less programmer oriented, and for them confusion of mixing two expression kinds could be even worse.
This is my humble opinion/feeling only and I understand others might feel different. 

In any case, you need to document what the Elsa "Javascript" (quotes intentional since it is not 100% javascript and is not 100% C#) can do and how it behaves. And that documentation should be targeted (as much as possible) to users who don't make their living writing code.

That being said I do understand that it could take a lot of work to make it happen.
But we should try to make "Dual Javascript" "blend" as much as we can.

Again pretending to be designer user only, the first thing I'd like to see is that all activity outputs (or models, don't know what is better name) are capable of doing `JSON.stringify(ActivityName.Output)`.

Why I chose this one? Well if you don't read the documentation, which ton of users don't even when it exists :), this would probably be the first step you would do when trying to discover what does the Activity  output look like :).
Moreover, since stringify does some number of things I believe that if we reach that some of the other JavaScript stuff that currently doesn't work would work too....

How to do that is another question. ....
A wild ideas, maybe the answer is forcing all activity outputs/models to implement some interface for which we know plays good with Jint. Haven't studied Jint in enough details to know if this is possible, or if it would "constraint" what model can be too much.
Maybe just documenting this "stringify" requirement so activity model developer is aware of it and can test it.......

Sorry for the longer post, let me know what you think :).....

**Comment by craigfowler** [comment]
Yes, I'm glad you started a new issue, and I agree, whilst one very specific problem might now be fixed, it does warrant a larger discussion about what we do in a more general manner.

Long posts also very welcome as far as I'm concerned, especially in an environment where we're not meeting face-to-face.  If you have ideas and don't write them down then they get lost.

I'll read in more depth in the future.  I _might even suggest converting this over to a discussion_ for the time being.  The nature of this kind of thing doesn't quite fit the concept of "a job which is either not started, in-progress or done".  Right now we're yet to decide exactly what we want to do about it.

**Comment by tomy2105** [comment]
Continuing on discussion from closed https://github.com/elsa-workflows/elsa-core/issues/761.

> ## Jint can't round trip it's own objects via stringify
> 
> Quite simply, if you use Jint to _create_ a plain old JS object like `{ foo: 5, bar: 'A string'}` then when you get that object back into the .NET world it will be a `System.Dynamic.ExpandoObject`. I don't see any other way to sensibly do that actually, since in JS you can just creates objects without needing to instantiate classes.

OK good to know you opened issue in Jint however I'm a little bit confused with your statement :).
What Jint returns, unless I'm reading code completely wrong, is `JSValue` object. And not `System.Dynamic.ExpandoObject`. If I missed something here please correct me :). And I'm pretty sure if you put this object "back" via `engine.SetValue` it would work with whole engine including `JSON.stringify`. 

Inside Elsa code there is conversion from `JSValue` to desired type in `Elsa.Scripting.JavaScript.Services.JavaScriptService`. Actually it first converts to object using Jints `ToObject` and then converts this object to desired type (either via converter if possible or via JSON encode/decode).

Maybe before `engine.SetValue` Elsa should do the opposite conversion? There seems to be Jint `FromObject` which returns `JSValue` for the opposite of `ToObject`. Don't know if conversions and/or JSON encode/decode is possible to do the opposite?

And if we manage to somehow solve this JSON parsed object to JSValue conversion we would automatically get the benefit of HTTP endpoint and request activity outputs work better when they are created from JSON.

What do you think?
